### PR TITLE
Add contains_content_items endpoint

### PR DESCRIPTION
**Pull request by bbaker6225**
## Ticket Link

Link to the associated ticket
[ENT-2563](https://openedx.atlassian.net/browse/ENT-2563)

## Post-review

Squash commits into discrete sets of changes

### CODE REVIEW

File: enterprise_catalog/apps/catalog/models.py [hunk 1]
```diff
@@ -116,6 +117,50 @@ def __str__(self):
...
+        Arguments:
+            content_id_values (set): Return only the content IDs that are members of this set
+                                      and members of the catalog.
+
+        Returns:
+            set: The set of content IDs that are members of the catalog.
+        """
+        all_metadata = self.catalog_query.contentmetadata_set.all()
+        return {metadata_chunk.content_key for metadata_chunk in all_metadata
+                if metadata_chunk.content_key in content_id_values}
```
Comment (bbaker6225) [inline enterprise_catalog/apps/catalog/models.py]: If we were to add `or metadata_chunk.parent_content_key in content_id_values` to this set comprehension, then we would be filtering to the content keys and their children that are in the catalog

File: enterprise_catalog/apps/api/v1/decorators.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+from functools import wraps
```
Comment (bbaker6225) [inline enterprise_catalog/apps/api/v1/decorators.py:1]: This file is just copied over from `edx-enterprise`

File: enterprise_catalog/apps/api/v1/views.py [hunk 3]
```diff
@@ -16,15 +23,30 @@
...
-        if action == 'create':
+        request_action = getattr(self, 'action', None)
+        if request_action == 'create':
             return EnterpriseCatalogCreateSerializer
 
         return EnterpriseCatalogSerializer
+
+    @method_decorator(require_at_least_one_query_parameter('course_run_ids', 'program_uuids'))
+    @action(detail=True)
+    def contains_content_items(self, request, uuid, course_run_ids, program_uuids):
```
Comment (bbaker6225) [inline enterprise_catalog/apps/api/v1/views.py:40]: It's annoying to still support both query params even though they both have the same effect now for backwards compatibility, but oh well

File: enterprise_catalog/apps/api/v1/tests/test_views.py [hunk 4]
```diff
@@ -157,9 +158,66 @@ def test_post(self):
...
+        """
+        response = self.client.get(self._get_contains_content_base_url(self.enterprise_catalog))
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_contains_content_items_no_catalog_query(self):
+        """
+        Verify the contains_content_items endpoint returns false if there is no associated catalog query
+        """
+        no_catalog_query_catalog = EnterpriseCatalogFactory(catalog_query=None)
+        url = self._get_contains_content_base_url(no_catalog_query_catalog) + '?program_uuids=test-uuid'
```
Comment (christopappas) [inline enterprise_catalog/apps/api/v1/tests/test_views.py:193]: For building your URLs, you can hand reverse args or kwargs as well to avoid string concatenation. Something like:

`reverse('api:v1:enterprise-catalog-contains-content-items', kwargs={'program_uuids': 'test-uuid'})`

(https://docs.djangoproject.com/en/3.0/ref/urlresolvers/#reverse)

File: enterprise_catalog/apps/api/v1/tests/test_views.py [hunk 5]
```diff
@@ -157,9 +158,66 @@ def test_post(self):
...
+        response = self.client.get(url)
```
Comment (christopappas) [inline enterprise_catalog/apps/api/v1/tests/test_views.py]: What type of object is response in this case?

File: enterprise_catalog/apps/api/v1/views.py [hunk 6]
```diff
@@ -16,15 +23,30 @@
...
-    serializer_class = EnterpriseCatalogSerializer
     queryset = EnterpriseCatalog.objects.all().order_by('created')
     permission_classes = [permissions.IsAuthenticated, permissions.IsAdminUser]
     authentication_classes = [JwtAuthentication, BearerAuthentication, SessionAuthentication]
     lookup_field = 'uuid'
 
     def get_serializer_class(self):
-        action = getattr(self, 'action', None)
-        if action == 'create':
+        request_action = getattr(self, 'action', None)
```
Comment (christopappas) [inline enterprise_catalog/apps/api/v1/views.py:32]: Curious about why the change from `action` to `request_action`, as `action` tends to have meaning within DRF itself (https://www.django-rest-framework.org/api-guide/viewsets/#viewset-actions)

File: enterprise_catalog/apps/api/v1/tests/test_views.py (hunk 5 above)
Comment (christopappas) [reply enterprise_catalog/apps/api/v1/tests/test_views.py]: I read a little further. Probably personal preference but to me scanning for a substring feels less intuitive than using a response's json() method `assert response.json()['contains_content_items'] is False`

File: enterprise_catalog/apps/catalog/models.py [hunk 7]
```diff
@@ -116,6 +116,32 @@ def __str__(self):
...
+        for content_key in content_keys:
+            try:
+                parent_content_key = ContentMetadata.objects.get(content_key=content_key).parent_content_key
+            except ContentMetadata.DoesNotExist:
+                parent_content_key = None
+            # The content key is contained in the catalog if its key is explictly part of the associated metadata, or
+            # its parent's key is.
+            contained_in_catalog = contained_in_catalog and (content_key in associated_metadata_content_keys
+                or parent_content_key in associated_metadata_content_keys)
+
```
Comment (christopappas) [inline enterprise_catalog/apps/catalog/models.py:147]: How long do we think content_keys can get? Would it make sense to break/return early if `contained_in_catalog` is False?

File: enterprise_catalog/apps/api/v1/views.py (hunk 3 above)
Comment (christopappas) [reply enterprise_catalog/apps/api/v1/views.py:40]: Do we need to worry about quoting/unquoting program uuids like we do course run ids? The answer is probably no, but it jumped out at me seeing course_run_ids being processed to some extent.

File: enterprise_catalog/apps/catalog/models.py (hunk 7 above)
Comment (bbaker6225) [reply enterprise_catalog/apps/catalog/models.py:147]: Sure, I can make it break early

File: enterprise_catalog/apps/api/v1/views.py (hunk 6 above)
Comment (bbaker6225) [reply enterprise_catalog/apps/api/v1/views.py:32]: quality was complaining about redefining `action` (imported from drf)

Comment (christopappas) [reply enterprise_catalog/apps/api/v1/views.py:32]: okay 🤦‍♂

File: enterprise_catalog/apps/api/v1/tests/test_views.py (hunk 4 above)
Comment (bbaker6225) [reply enterprise_catalog/apps/api/v1/tests/test_views.py:193]: Unfortunately I'm pretty sure `kwargs` can't be used for query params (or at least I have not figured out how it can)

File: enterprise_catalog/apps/api/v1/views.py (hunk 3 above)
Comment (bbaker6225) [reply enterprise_catalog/apps/api/v1/views.py:40]: Nah the dashes/hyphens in program uuids work fine for me in testing, as opposed to the pluses in course run ids.
Mostly copied from `edx-enterprise` also: https://github.com/edx/edx-enterprise/blob/30a3a930358e059fa469fdfadd1ddbb084cb29a1/enterprise/api/v1/views.py#L332-L333

File: enterprise_catalog/apps/api/v1/tests/test_views.py [hunk 8]
```diff
@@ -157,9 +158,72 @@ def test_post(self):
...
+
+    def _get_contains_content_base_url(self, enterprise_catalog):
+        """
+        Helper to construct the base url for the contains_content_items endpoint
+        """
+        return reverse('api:v1:enterprise-catalog-contains-content-items', kwargs={'uuid': enterprise_catalog.uuid})
+
+    def _assert_correct_response(self, url, expected_value):
+        """
+        Helper to asssert that the contains_content_items endpoint specified by the url returns the correct value
```
Comment (adamstankiewicz) [inline enterprise_catalog/apps/api/v1/tests/test_views.py]: nit: little typo in the docstring -- "assert"

File: enterprise_catalog/apps/api/v1/tests/test_views.py [hunk 9]
```diff
@@ -157,9 +158,72 @@ def test_post(self):
...
+        url = self._get_contains_content_base_url(self.enterprise_catalog) + '?course_run_ids=' + content_key
+        self._assert_correct_response(url, True)
+
+    def test_contains_content_items_parent_keys_in_catalog(self):
+        """
+        Verify the contains_content_items endpoint returns True if the parent's key is in the catalog
+        """
+        child_key = 'child-key'
+        parent_metadata = ContentMetadataFactory(content_key='parent-key')
+        ContentMetadataFactory(content_key='child-key', parent_content_key=parent_metadata.content_key)
```
Comment (adamstankiewicz) [inline enterprise_catalog/apps/api/v1/tests/test_views.py]: nit: would it make sense to use `child_key` variable instead of hardcoded string here?

File: enterprise_catalog/apps/api/v1/views.py [hunk 10]
```diff
@@ -16,15 +23,30 @@
...
+        if request_action == 'create':
             return EnterpriseCatalogCreateSerializer
 
         return EnterpriseCatalogSerializer
+
+    @method_decorator(require_at_least_one_query_parameter('course_run_ids', 'program_uuids'))
+    @action(detail=True)
```
Comment (adamstankiewicz) [inline enterprise_catalog/apps/api/v1/views.py:39]: Is there any difference between the `@action` decorator and the `@detail_route` decorator? Looks like we use the [`@detail_route` decorator in edx-enterprise](https://github.com/edx/edx-enterprise/blob/23af86a57a944da2839ad50d348adc403158fa13/enterprise/api/v1/views.py#L139).

Comment (bbaker6225) [reply enterprise_catalog/apps/api/v1/views.py:39]: `@action` is the newer version
https://www.django-rest-framework.org/community/3.9-announcement/#action-decorator-replaces-list_route-and-detail_route

File: enterprise_catalog/apps/api/v1/tests/test_views.py (hunk 9 above)
Comment (bbaker6225) [reply enterprise_catalog/apps/api/v1/tests/test_views.py]: yes it would

**Review by christopappas — APPROVED** [review]
